framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -375,9 +375,111 @@ public void processClassTree(ClassTree classTree) {
 
         checkExtendsImplements(classTree);
 
+        checkQualifierParam(classTree);
+
         super.visitClass(classTree, null);
     }
 
+    /**
+     * Issues an error if {@code classTree} has polymorphic fields but is not annotated with
+     * {@code @HasQualifierParameter}. Always issue a warning if the type of a static field is
+     * annotated with a polymorphic qualifier.
+     *
+     * <p>Issues an error if (@code classTree} extends or implements a class/interface that has a
+     * qualifier parameter, but this class does not.
+     *
+     * @param classTree the ClassTree to check foor polymorphic fields
+     */
+    private void checkQualifierParam(ClassTree classTree) {
+        Set<AnnotationMirror> polyWithOutQualiferParam = AnnotationUtils.createAnnotationSet();
+        Set<AnnotationMirror> polys = AnnotationUtils.createAnnotationSet();
+
+        boolean conflictingErrorReported = false;
+        for (AnnotationMirror top : atypeFactory.getQualifierHierarchy().getTopAnnotations()) {
+            TypeElement classElement = TreeUtils.elementFromDeclaration(classTree);
+            AnnotationMirror poly =
+                    atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+            if (poly != null) {
+                polys.add(poly);
+            }
+
+            if (!conflictingErrorReported
+                    && atypeFactory.hasExplicitQualifierParameterInHierarchy(classElement, top)
+                    && atypeFactory.hasExplicitNoQualifierParameterInHierarchy(classElement, top)) {
+                checker.report(Result.failure("conflicting.qual.param", top), classTree);
+                conflictingErrorReported = true;
```

Comment:
Why does this report only the first conflicting top annotations? If there are multiple annotations on a class, wouldn't it be useful to highlight all top annotations that have conflicts?